README
0. Legal Stuff
--------------
 The Code included in this archive is Copyright (c) 2004-2007
 Ove Soerensen, Portions Copyright (c) 2006 Johannes Hofmann & 2007 Stephen
 M. Rumble

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions
  are met:
  1. Redistributions of source code must retain the above copyright
     notice, this list of conditions and the following disclaimer. 
  2. Redistributions in binary form must reproduce the above copyright
     notice, this list of conditions and the following disclaimer in the
     documentation and/or other materials provided with the distribution.

  THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS 'AS IS' AND
  ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
  ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE   
  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS   
  OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)     
  HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY 
  OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF    
  SUCH DAMAGE.

1. Who?
-------
 estd is written by Ove Soerensen. 2009 I have taken over maintainership.
 You can reach me via email (Johannes.Hofmann@gmx.de). Updates and additional
 information might be available at my website:
 http://www.ecademix.com/JohannesHofmann/estd.html

2. What?
--------
 This daemon dynamically sets the CPU-frequency on SpeedStep-,
 PowerNow-, and ACPI P-States enabled CPUs depending on current cpu-utilization. 
 It supports NetBSD 3.0 or later and DragonFly BSD.

3. How?
-------
 Installation should be really easy. Just untar, make and make install as root.
 Make sure you have support for Enhanced SpeedStep / PowerNow, or ACPI P-States
 enabled in your kernel-config and you are running as root before starting the
 daemon. You can simply run the daemon without further arguments, it will pick
 some sensible defaults for you (but it won't fork by default). For command line
 options and further details please check the man page.


ChangeLog
estd ChangeLog
==============

estd-r11
* Fix incorrect static array initialization introduced in r10.
  submitted-by: Leonardo Taccari <iamleot@gmail.com>

estd-r10
* Fix build on DragonFly due to pidfile changes.
* Add support for TECH_LOONGSON, TECH_ROCKCHIP, and TECH_GENERIC.
  Taken from pkgsrc, submitted-by: Leonardo Taccari <iamleot@gmail.com>

estd-r9
* Improve SMP support on NetBSD.
  Submitted by: vincent@labri.fr

estd-r8
* Add Intrepid support.
  Submitted by: Michael <macallan@netbsd.org>

estd-r7
* Add support for ACPI P-States on DragonFly.
* Improve multi-core support on DragonFly. For each CPU domain with individually
  controllable CPU frequency, the load of the most loaded CPU is used instead of
  the overall avarage load.
  This avoids issues on SMP systems where a single-threaded CPU bound process
  will lead to at most 50% overall load.
* Ove Soerensen has asked me to take over maintainership of estd.
  The new project page is at http://www.ecademix.com/JohannesHofmann/estd.html


netbsd_envstat_temp.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <time.h>
#include <regex.h>
#include <prop/proplib.h>

#ifndef PATH_ENVSTAT
#define PATH_ENVSTAT "/usr/sbin/envstat"
#endif

static void
prop_iterate0(void (*cbfunc)(void *, const char *, prop_object_t), char *cbarg, char *key, size_t keylen, prop_object_t parent, prop_object_t obj)
{
	prop_object_iterator_t itr;
	prop_object_t obj2;
	const char *p;
	char *k;

	switch (prop_object_type(obj)) {
	case PROP_TYPE_BOOL:
	case PROP_TYPE_NUMBER:
	case PROP_TYPE_STRING:
	case PROP_TYPE_DATA:
		(*cbfunc)(cbarg, key, obj);
		break;

	case PROP_TYPE_ARRAY:
		itr = prop_array_iterator(obj);
		for (obj2 = prop_object_iterator_next(itr); obj2 != NULL;
		    obj2 = prop_object_iterator_next(itr)) {
			prop_iterate0(cbfunc, cbarg, key, keylen, obj, obj2);
		}
		prop_object_iterator_release(itr);
		break;
	case PROP_TYPE_DICTIONARY:
		itr = prop_dictionary_iterator(obj);
		for (obj2 = prop_object_iterator_next(itr); obj2 != NULL;
		    obj2 = prop_object_iterator_next(itr)) {
			prop_iterate0(cbfunc, cbarg, key, keylen, obj, obj2);
		}
		prop_object_iterator_release(itr);
		break;
	case PROP_TYPE_DICT_KEYSYM:
		p = prop_dictionary_keysym_cstring_nocopy(obj);
		k = key + strlen(key);
		if (key[0] != '\0')
			strlcat(key, ".", keylen);
		strlcat(key, p, keylen);
		obj2 = prop_dictionary_get_keysym(parent, obj);
		prop_iterate0(cbfunc, cbarg, key, keylen, obj, obj2);
		*k = '\0';
		break;
	default:
		fprintf(stderr, "estd: UNKNOWN PROP TYPE\n");
		break;
	}
}

static void
prop_iterate(void (*cbfunc)(void *, const char *, prop_object_t), char *cbarg, prop_object_t obj)
{
	char *key;
	size_t keylen = 1024;
	key = alloca(keylen);
	key[0] = '\0';
	prop_iterate0(cbfunc, cbarg, key, keylen, NULL, obj);
}


char *
freadin(FILE *fh)
{
#define READIN_BLOCKSIZE	(1024 * 8)
	char *buf, *p;
	size_t size, done, rc;

	size = READIN_BLOCKSIZE;
	buf = p = malloc(size);
	if (buf == NULL)
		return NULL;
	done = 0;
	for (;;) {
		rc = fread(p, 1, READIN_BLOCKSIZE, fh);
		if (rc < 0) {
			fprintf(stderr, "estd: fread: %s\n", strerror(errno));
			free(buf);
			return NULL;
		}
		done += rc;
		if (rc < READIN_BLOCKSIZE)
			break;

		size += READIN_BLOCKSIZE;
		p = realloc(buf, size);
		if (p == NULL) {
			fprintf(stderr, "estd: malloc: %s\n", strerror(errno));
			free(buf);
			return NULL;
		}
		buf = p;
		p = buf + done;
	}

	if (done == size) {
		p = realloc(buf, size + 1);
		if (p == NULL) {
			fprintf(stderr, "estd: malloc: %s\n", strerror(errno));
			free(buf);
			return NULL;
		}
		buf = p;
	}
	p = buf + done;
	*p = '\0';

	return buf;
}

struct keychecker {
	regex_t re;
	double limit;
	int result;
};

static void
prop_check_callback(void *arg, const char *key, prop_object_t obj)
{
	struct keychecker *keychecker;
	regmatch_t re_pmatch[1];
	double celsius;
	uint64_t num;

	keychecker = (struct keychecker *)arg;

	re_pmatch[0].rm_so = 0;
	re_pmatch[0].rm_eo = strlen(key);
	if (regexec(&keychecker->re, key, 1, re_pmatch, REG_STARTEND) != 0)
		return;

	switch (prop_object_type(obj)) {
	case PROP_TYPE_BOOL:
#ifdef NETBSD_ENVSTAT_DEBUG
		fprintf(stderr, "%s=%s\n", key, prop_bool_true(obj) ? "true" : "false");
#endif
		break;
	case PROP_TYPE_NUMBER:
		num = prop_number_unsigned_integer_value(obj);
		num -= 273150000;
		celsius = num / 1000000;
		if (celsius >= keychecker->limit)
			keychecker->result = 1;
		break;
	case PROP_TYPE_STRING:
#ifdef NETBSD_ENVSTAT_DEBUG
		{
			const char *p;
			p = prop_string_cstring_nocopy(obj);
			fprintf(stderr, "%s=\"%s\"\n", key, p);
		}
#endif
		break;
	case PROP_TYPE_DATA:
#ifdef NETBSD_ENVSTAT_DEBUG
		fprintf(stderr, "%s=<PROP_TYPE_DATA>\n", key);
#endif
		break;
	default:
#ifdef NETBSD_ENVSTAT_DEBUG
		fprintf(stderr, "%s: UNKNOWN PROP TYPE\n", key);
#endif
		break;
	}
}

static int
check_overheat(const char *device, double limit)
{
	prop_object_t propobj;
	struct keychecker keychecker;
	FILE *fh;
	int rc;
	char pattern[128];
	char *xml;

	/* exec "envstat -x" and parse */
	fh = popen(PATH_ENVSTAT " -x", "r");
	if (fh == NULL) {
		fprintf(stderr, "popen: envstat: %s\n", strerror(errno));
		return -1;
	}
	xml = freadin(fh);
	pclose(fh);

	if (xml == NULL)
		return -1;

	propobj = prop_dictionary_internalize(xml);
	free(xml);
	if (propobj == NULL) {
		fprintf(stderr, "estd: cannot read result of `envstat -x`\n");
		return -1;
	}

	memset(&keychecker, 0, sizeof(keychecker));
	snprintf(pattern, sizeof(pattern), "%s\\.cur-value", device);

	rc = regcomp(&keychecker.re, pattern, REG_EXTENDED|REG_ICASE);
	keychecker.limit = limit;

	prop_iterate(prop_check_callback, (void *)&keychecker, propobj);
	prop_object_release(propobj);

	regfree(&keychecker.re);

	return keychecker.result;
}

int
is_overheat(const char *device, double celsius, unsigned int cachetime)
{
	static struct timespec ts_last, ts_now;
	static int result;

	if (device == NULL)
		return 0;

	clock_gettime(CLOCK_MONOTONIC, &ts_now);
	if ((ts_now.tv_sec - ts_last.tv_sec) > cachetime) {
		result = check_overheat(device, celsius);
		ts_last = ts_now;
	}

	return result;
}

#if 0
int
main(int argc, char *argv[])
{
	int fire;

	for (;;) {
		fire = check_overheat("coretemp[0-9]+", 59.0);
		printf("overheat=%d\n", fire);
		fflush(stdout);

		sleep(1);
	}
}
#endif


estd.c
/*
 * Enhanced Speedstep & PowerNow Daemon for NetBSD & DragonFly BSD, Code (c) 
 * 2004-2007 by Ove Soerensen, Portions (c) 2006,2009 Johannes Hofmann, 2007
 * Stephen M. Rumble
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met: 1. Redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer. 2.
 * Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS 'AS IS' AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE   FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 */


#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <time.h>
#include <sys/param.h>
#include <sys/sysctl.h>
#if defined(__DragonFly__)
 #include <kinfo.h>
 #include <libutil.h>
#else
 #include <sys/sched.h>
 #include <util.h>
#endif
#include <errno.h>
#include <signal.h>

#define ESTD_VERSION "Release 11"
#define BATTERY 0
#define SMOOTH 1
#define AGGRESSIVE 2
#define DEF_POLL 500000
#define MIN_POLL 10000
#define DEF_HIGH 80
#define DEF_LOW 40
#define TECH_UNKNOWN 0
#define TECH_EST 1
#define TECH_POWERNOW 2
#define TECH_ACPI 3
#define TECH_INTREPID 4
#define TECH_LOONGSON 5
#define TECH_ROCKCHIP 6
#define TECH_GENERIC 7
#define TECH_MAX 7
 
/* this is ugly, but... <shrug> */
#define MAX_FREQS 32
#define SYSCTLBUF 255
#define MAX_CPUS  128

extern char    *optarg;
extern int      optind;

#if defined(__DragonFly__)
 #define useconds_t unsigned int
#endif

/* command-line options */
int             daemonize = 0;
int             verbose = 0;
int             nicemod = 0;
int             strategy = SMOOTH;
useconds_t      poll = DEF_POLL;
int             high = DEF_HIGH;
int             low = DEF_LOW;
useconds_t      lowgrace = 0;
int             minmhz = 0;
int             maxmhz = INT_MAX;
int             listfreq = 0;
int             tech = TECH_UNKNOWN;
int             use_clockmod = 0;
int             clockmod_min = -1;
int             clockmod_max = -1;
#ifdef OVERHEAT_HACK
extern int is_overheat(const char *, double, unsigned int);
#define DEF_SENSORPOLL	15	/* check interval is 15 seconds */
const char      *sensordev;
unsigned int    sensorpoll = DEF_SENSORPOLL;
double          sensorcelsius = 90.0;	/* default */
#endif

/* a domain is a set of CPUs for which the frequency must be set together */
struct domain {
	int         *cpus;
	int          ncpus;
	char        *freqctl;
	char        *setctl;
	useconds_t   lowtime;
	int          freqtab[MAX_FREQS];
	int          nfreqs;
	int          minidx;
	int          maxidx;
	int          curcpu;
	int          curfreq;
};

int             ncpus = 0;
struct domain  *domain;
int             ndomains;

#if defined(__DragonFly__)
 static struct kinfo_cputime *cp_time;
 static struct kinfo_cputime *cp_old;
 size_t cp_time_len;
#else
 static int cpumib[2] = {CTL_KERN, KERN_CP_TIME};
 static u_int64_t cp_time[MAX_CPUS][CPUSTATES];
 static u_int64_t cp_old[MAX_CPUS][CPUSTATES];
 static u_int64_t cp_diff[MAX_CPUS][CPUSTATES];
 static const size_t cp_time_max_size = sizeof(cp_time[0]) * MAX_CPUS;
#endif

static char	*techdesc[TECH_MAX + 1] = {"Unknown",
				"Enhanced SpeedStep",
				"PowerNow",
				"ACPI P-States",
				"Intrepid",
				"Loongson",
				"Rockchip",
				"Generic"
				};
static char	*freqctl[TECH_MAX + 1] = {	"",	
				"machdep.est.frequency.available",
				"machdep.powernow.frequency.available",
				"hw.acpi.cpu.px_dom0.available",
				"machdep.intrepid.frequency.available",
				"machdep.loongson.frequency.available",
				"machdep.cpu.frequency.available",
				"machdep.frequency.available"
				};
static char	*setctl[TECH_MAX + 1] = {	"",
				"machdep.est.frequency.target",
				"machdep.powernow.frequency.target",
				"hw.acpi.cpu.px_dom0.select",
				"machdep.intrepid.frequency.target",
				"machdep.loongson.frequency.target",
				"machdep.cpu.frequency.target",
				"machdep.frequency.current"
				};

void
usage()
{
	printf("usage: estd [-d] [-o] [-n] [-A] [-C] [-E] [-I] [-L] [-R] [-P] [-G] [-a] [-s] [-b] [-p poll interval in us] [-g grace period] [-l low watermark percentage] [-h high watermark percentage] [-m minimum MHz] [-M maximum MHz]\n");
	printf("       estd -v\n");
	printf("       estd -f\n");
	exit(1);
}

void
version()
{
	printf("estd ");
	printf(ESTD_VERSION);
	printf(" Copyright (c) 2004-2009 Ove Soerensen\n");
	printf("Details at http://www.ecademix.com/JohannesHofmann/estd.html - Contact via Johannes.Hofmann@gmx.de\n");
	exit(0);
}

void *
ecalloc(size_t number, size_t size) {
	void *ret = calloc(number, size);
	if (ret == NULL) {
		fprintf(stderr, "estd: calloc failed (errno %d)\n", errno);
		exit(1);
	}

	return ret;
}

int
acpi_init_domain(int d)
{
	char name[256];
	char members[SYSCTLBUF];
	char *mp;
	size_t memberssize = SYSCTLBUF;
	int i;

	snprintf(name, sizeof(name), "hw.acpi.cpu.px_dom%d.members", d);
	if (sysctlbyname(name, &members, &memberssize, NULL, 0) < 0)
		return 1;

	ndomains = d + 1;
	domain = realloc(domain, ndomains * sizeof(struct domain));
	if (domain == NULL) {
		fprintf(stderr, "estd: realloc failed (errno %d)\n", errno);
		exit(1);
	}
	memset(&domain[d], 0, sizeof(struct domain));

	domain[d].ncpus = 0;
	domain[d].cpus = ecalloc(ncpus, sizeof(int));
	mp = members;
	while ((mp = strstr(mp, "cpu"))) {
		mp += strlen("cpu");
		domain[d].cpus[domain[d].ncpus++] = strtol(mp, &mp, 10);
	}

	asprintf(&domain[d].freqctl, "hw.acpi.cpu.px_dom%d.available", d);
	asprintf(&domain[d].setctl, "hw.acpi.cpu.px_dom%d.select", d);
	if (domain[d].setctl == NULL || domain[d].freqctl == NULL) {
		fprintf(stderr, "estd: asprintf failed\n");
		exit(1);
	}

	if ((!daemonize) && (verbose))
		for (i = 0; i < domain[d].ncpus; i++)
			printf("estd: domain %d: member %d\n", d, domain[d].cpus[i]);

	return 0;
}

int
acpi_init()
{
	int d = 0;

	while (acpi_init_domain(d) == 0)
		d++;

	return d > 0 ? 0 : 1;
}


/* returns cpu-usage in percent, mean over the sleep-interval or -1 if an error occured */
#if defined(__DragonFly__)
int
get_cputime()
{
	size_t len = cp_time_len;

	memcpy(cp_old, cp_time, cp_time_len);

	if (sysctlbyname("kern.cputime", cp_time, &len, NULL, 0) < 0) {
		fprintf(stderr, "estd: Cannot get CPU status\n");
		exit(1);	
	}

	return 0;
}

/* get maximum load of a cpu in domain d */
int
get_cpuusage(int d)
{
	int                  i, cpu, load, max_load = 0;
	u_int64_t            total_time;
	struct kinfo_cputime cp_diff;

	for (i = 0; i < domain[d].ncpus; i++) {
		cpu = domain[d].cpus[i];
		total_time = 0;

		cp_diff.cp_user = cp_time[cpu].cp_user - cp_old[cpu].cp_user;
		total_time += cp_diff.cp_user;
		cp_diff.cp_nice = cp_time[cpu].cp_nice - cp_old[cpu].cp_nice;
		total_time += cp_diff.cp_nice;
		cp_diff.cp_sys = cp_time[cpu].cp_sys - cp_old[cpu].cp_sys;
		total_time += cp_diff.cp_sys;
		cp_diff.cp_intr = cp_time[cpu].cp_intr - cp_old[cpu].cp_intr;
		total_time += cp_diff.cp_intr;
		cp_diff.cp_idle = cp_time[cpu].cp_idle - cp_old[cpu].cp_idle;
		total_time += cp_diff.cp_idle;

		if (total_time > 0) {
			load = 100 - ((cp_diff.cp_idle + (cp_diff.cp_nice * nicemod)) * 100) / total_time;
			if (load > max_load)
				max_load = load;
		}
	}

	return max_load;
}
#else

int
get_cputime(void)
{
	size_t cp_time_size = cp_time_max_size;

	memcpy(cp_old, cp_time, cp_time_max_size);
	if (sysctl(cpumib, 2, &cp_time, &cp_time_size, NULL, 0) < 0) {
		fprintf(stderr, "estd: Cannot get CPU status\n");
		exit(1);
	}

	ncpus = cp_time_size / sizeof cp_time[0];

	return 0;
}

int
get_cpuusage(int d)
{
	u_int64_t	max_total_time = 0;
	int		cpu_of_max = 0;
	int		cpu;
	int             i;

	for (cpu = 0; cpu < ncpus; cpu++) {
		u_int64_t total_time = 0;

		for (i = 0; i < CPUSTATES; i++) {
			cp_diff[cpu][i] = cp_time[cpu][i] - cp_old[cpu][i];
			if (i != CP_IDLE)
				total_time += cp_diff[cpu][i];
		}
		if (total_time > max_total_time) {
			max_total_time = total_time;
			cpu_of_max = cpu;
		}
	}

	max_total_time += cp_diff[cpu_of_max][CP_IDLE];

	/* we've probably been interrupted by a signal... */
	if (max_total_time < 1) return -1; 

	return (100 - ((cp_diff[cpu_of_max][CP_IDLE] +
			(cp_diff[cpu_of_max][CP_NICE] * nicemod)) * 100) /
				max_total_time);
}
#endif

/* sets the cpu frequency */
void
set_freq(int d)
{
	int freq = domain[d].freqtab[domain[d].curfreq];

	if ((!daemonize) && (verbose))
		printf("%i MHz\n", freq);
	if (sysctlbyname(domain[d].setctl, NULL, NULL, &freq, sizeof(freq)) < 0) {
		fprintf(stderr, "estd: Cannot set CPU frequency (maybe you aren't root?)\n");
		exit(1);
	}
}

void
set_clockmod(int level)
{
#if !defined(__DragonFly__)
	if (!use_clockmod || level == -1)
		return;

	if ((!daemonize) && (verbose))
		printf("clockmod level: %i\n", level);
	if (sysctlbyname("machdep.clockmod.target", NULL, NULL, &level,
	    sizeof(level)) < 0) {
		fprintf(stderr, "estd: Cannot set clockmod level (maybe you aren't root?)\n");
		exit(1);
	}
#endif
}

/* need this callback for sorting the frequency-list */
int
freqcmp(const void *x, const void *y)
{
	return *((int *) x) - *((int *) y);
}

/* clean up the pidfile and clockmod on exit */
void
sighandler(int sig)
{
	set_clockmod(clockmod_max);
	exit(0);
}

/* switch strategy on SIGUSR{1,2} */
void
sigusrhandler(int sig)
{
	switch (sig) {
		case SIGUSR1:
				if (strategy>BATTERY) strategy--;
				break;
		case SIGUSR2:
				if (strategy<AGGRESSIVE) strategy++;
				break;
	}
}

int
main(int argc, char *argv[])
{
	int             ch;
	int             i;
	char            frequencies[SYSCTLBUF];	/* XXX Ugly */
	char           *fp;
	size_t          freqsize = SYSCTLBUF;
	int	            curstrat = strategy;
	int             d;
	FILE           *fexists;
#if defined(__DragonFly__)
	struct pidfh   *pdf;
#endif

	/* get command-line options */
#ifdef OVERHEAT_HACK
	while ((ch = getopt(argc, argv, "vfdonACEGILPT:t:asbp:h:l:g:m:M:c:")) != -1)
#else
	while ((ch = getopt(argc, argv, "vfdonACEGILPasbp:h:l:g:m:M:")) != -1)
#endif
		switch (ch) {
		case 'v':
			version();
			/* NOTREACHED */
		case 'f':
			listfreq = 1;
			break;
		case 'd':
			daemonize = 1;
			break;
		case 'o':
			verbose = 1;
			break;
		case 'n':
			nicemod = 1;
			break;
		case 'A':
			tech = TECH_ACPI;
			break;
		case 'C':
			#if !defined(__DragonFly__)
			 use_clockmod = 1;
			 break;
			#else
			 fprintf(stderr, "-C not available under DragonFly\n");
			 exit(1);
			#endif
		case 'E':
			tech = TECH_EST;
			break;
		case 'G':
			tech = TECH_GENERIC;
			break;
		case 'I':
			tech = TECH_INTREPID;
			break;
                case 'L':
                        tech = TECH_LOONGSON;
                        break;
		case 'P':
			tech = TECH_POWERNOW;
			break;
		case 'R':
			tech = TECH_ROCKCHIP;
			break;
		case 'a':
			strategy = AGGRESSIVE;
			break;
		case 's':
			strategy = SMOOTH;
			break;
		case 'b':
			strategy = BATTERY;
			break;
		case 'p':
			poll = atoi(optarg);
			break;
		case 'h':
			high = atoi(optarg);
			break;
		case 'l':
			low = atoi(optarg);
			break;
		case 'g':
			lowgrace = atoi(optarg);
			break;
		case 'm':
			minmhz = atoi(optarg);
			break;
		case 'M':
			maxmhz = atoi(optarg);
			break;
#ifdef OVERHEAT_HACK
		case 'T':
			sensordev = optarg;
			break;
		case 't':
			sensorpoll = atoi(optarg);
			break;
		case 'c':
			sensorcelsius = atof(optarg);
			break;
#endif
		default:
			usage();
			/* NOTREACHED */
		}

	ndomains = 1;
	domain = ecalloc(ndomains, sizeof(struct domain));

#if defined(__NetBSD__) || defined(__DragonFly__)
# if defined(__DragonFly__)
	if (kinfo_get_cpus(&ncpus)) {
		fprintf(stderr, "estd: Cannot get number of cpus\n");
		exit(1);
	}
	cp_time = ecalloc(ncpus, sizeof(struct kinfo_cputime));
	cp_old  = ecalloc(ncpus, sizeof(struct kinfo_cputime));
	cp_time_len = ncpus * sizeof(struct kinfo_cputime);
# elif defined(__NetBSD__)
	size_t ncpus_len = sizeof(ncpus);
	if (sysctlbyname("hw.ncpu", &ncpus, &ncpus_len, NULL, 0) != 0) {
		fprintf(stderr, "estd: Cannot get number of cpus\n");
		exit(1);
	}
# endif
	domain[0].ncpus = ncpus;
	domain[0].cpus = ecalloc(ncpus, sizeof(int));
	for (i = 0; i < domain[0].ncpus; i++)
		domain[0].cpus[i] = i;
#endif

	/* try to guess cpu-scaling technology */
	if (tech == TECH_UNKNOWN) {
		for (tech = 1; tech <= TECH_MAX; tech++) {
			if (sysctlbyname(freqctl[tech], &frequencies, &freqsize, NULL, 0) >= 0) break;
		}
		if (tech > TECH_MAX) {
			fprintf(stderr, "estd: Cannot guess CPU-scaling technology. (maybe you are missing some kernel-option?)\n");
			exit(1);
		}
	}

	if (tech == TECH_ACPI) {
		if (acpi_init()) {
			fprintf(stderr, "estd: Cannot ACPI P-States\n");
			exit(1);
		}
	} else {
		domain[0].freqctl = freqctl[tech];
		domain[0].setctl = setctl[tech];
	}

	if ((high <= low) || (low < 0) || (low > 100) || (high < 0) || (high > 100)) {
		fprintf(stderr, "estd: Invalid high/low watermark combination\n");
		exit(1);
	}

	if (poll < MIN_POLL) {
		fprintf(stderr, "estd: Poll interval is too low (minimum %i)\n", MIN_POLL);
		exit(1);
	}

	if (minmhz > maxmhz) {
		fprintf(stderr, "estd: Invalid minimum/maximum MHz combination\n");
		exit(1);
	}

	/* for each cpu domain... */
	for (d = 0; d < ndomains; d++) {
		/* get supported frequencies... */
		if (sysctlbyname(domain[d].freqctl, &frequencies, &freqsize, NULL, 0) < 0) {
			fprintf(stderr, "estd: Cannot get supported frequencies (maybe you forced the wrong CPU-scaling technology?)\n");
			exit(1);
		}
		fp = &frequencies[0];
		while ((domain[d].nfreqs < MAX_FREQS) && ((domain[d].freqtab[domain[d].nfreqs++] = strtol(fp, &fp, 10)) != 0));
		domain[d].nfreqs--;
		if (domain[d].nfreqs <= 0) {
			fprintf(stderr, "estd: No supported frequencies found?! (please report this error)\n");
			exit(1);
		}
		/* ...and sort them in ascending order */
		qsort(&domain[d].freqtab, domain[d].nfreqs, sizeof(domain[d].freqtab[0]), &freqcmp);
		/* some sanity checks */
		while ((domain[d].minidx < domain[d].nfreqs) && (domain[d].freqtab[domain[d].minidx] < minmhz))
			domain[d].minidx++;
		if (domain[d].minidx >= domain[d].nfreqs) {
			fprintf(stderr, "estd: Minimum Frequency is too high\n");
			exit(1);
		}
		domain[d].maxidx = domain[d].nfreqs - 1;
		while ((domain[d].maxidx > -1) && (domain[d].freqtab[domain[d].maxidx] > maxmhz))
		domain[d].maxidx--;
		if (domain[d].maxidx < 0) {
			fprintf(stderr, "estd: Maximum Frequency is too low\n");
			exit(1);
		}
		if (domain[d].freqtab[domain[d].minidx] > domain[d].freqtab[domain[d].maxidx]) {
			fprintf(stderr, "estd: No supported frequency within given range found\n");
			exit(1);
		}
	}

	if (listfreq) {
		printf("Supported frequencies (%s Mode):\n",techdesc[tech]);
		for (d = 0; d < ndomains; d++) {
			printf("Domain %d:\n", d);
			for (i = 0; i < domain[d].nfreqs; i++) {
				printf("%i MHz\n", domain[d].freqtab[i]);
			}
		}
		exit(0);
	}

	#if !defined(__DragonFly__)
	{
		char   *lastfp = NULL;
		char	clockmods[SYSCTLBUF];
		size_t	len = sizeof(clockmods);

		if (sysctlbyname("machdep.clockmod.available", &clockmods, &len,
		    NULL, 0) >= 0) {
			fp = &clockmods[0];
			while (fp != lastfp) {
				lastfp = fp;
				i = strtol(fp, &fp, 10);
				if (i < clockmod_min || clockmod_min == -1)
					clockmod_min = i;
				if (i > clockmod_max || clockmod_max == -1)
					clockmod_max = i;
			}
		}
	}
	#endif

	if ((fexists = fopen("/var/run/estd.pid", "r")) != NULL) {
		fprintf(stderr, "estd: Pidfile /var/run/estd.pid exists, remove it if you are sure it shouldn't be there (maybe another instance of estd is already running?)\n");
		fclose(fexists);
		exit(1);
	}
	
	/* all ok, here we go */
	if (daemonize) {
		if (fork()) {
			if (verbose)
				printf("estd: Forked\n");
			exit(0);
		}
	} else {
		printf("estd: Not detaching from terminal\n");
	}

#if defined(__DragonFly__)
	pdf = pidfile_open(NULL, 600, NULL);
	if (pdf == NULL) {
		fprintf(stderr, "estd: Cannot write pidfile (maybe you aren't root?)\n");
		exit(1);
	} else {
		pidfile_write(pdf);
	}
#else
	if (pidfile(NULL)) {
		fprintf(stderr, "estd: Cannot write pidfile (maybe you aren't root?)\n");
		exit(1);
	}
#endif
	
	/* init some vars and set inital frequency */
	signal(SIGHUP, SIG_IGN);
	signal(SIGINT, &sighandler);
	signal(SIGPIPE, &sighandler);
	signal(SIGTERM, &sighandler);
	signal(SIGUSR1, &sigusrhandler);
	signal(SIGUSR2, &sigusrhandler);

	for (d = 0; d < ndomains; d++) {
		domain[d].curfreq = domain[d].minidx;
		set_freq(d);
	}
	set_clockmod(clockmod_min);

	/* the big processing loop, we will only exit via signal */
	while (1) {
#ifdef OVERHEAT_HACK
		int overheating = is_overheat(sensordev, sensorcelsius, sensorpoll);
#endif
		get_cputime();
		for (d = 0; d < ndomains; d++) {
			domain[d].curcpu = get_cpuusage(d);
			if ((!daemonize) && (verbose))
				printf("estd: load(%d) %d\n", d, domain[d].curcpu);
			if (domain[d].curcpu != -1) {
				/* strategy can change anytime (SIGUSR) */ 
				curstrat = strategy;
#ifdef OVERHEAT_HACK
				if (overheating) {
					if ((!daemonize) && (verbose))
						printf("estd: overheat\n");
					domain[d].curcpu = 0;	/* force down clock */
				}
#endif
				if ((domain[d].curfreq > domain[d].minidx) && (domain[d].curcpu < low)) {
					if (domain[d].lowtime < lowgrace)
						domain[d].lowtime += poll;

					if (domain[d].lowtime >= lowgrace) {
						if (curstrat == BATTERY)
							domain[d].curfreq = domain[d].minidx;
						else
							domain[d].curfreq--;
						set_freq(d);
						if (domain[d].curfreq == domain[d].minidx)
							set_clockmod(clockmod_min);
					}
				} else {
					domain[d].lowtime = 0;

					if ((domain[d].curfreq < domain[d].maxidx) && (domain[d].curcpu > high)) {
						if (curstrat == AGGRESSIVE)
							domain[d].curfreq = domain[d].maxidx;
						else
							domain[d].curfreq++;
						set_freq(d);
						if (domain[d].curfreq != domain[d].minidx)
							set_clockmod(clockmod_max);
					}
				}
			}
		}
		if (poll >= 1000000)
			sleep(poll / 1000000);
		usleep(poll % 1000000);
	}

	return 0;
}
